### PR TITLE
Support helm migration from current grafana/loki chart

### PR DESCRIPTION
**Pull request by trevorwhitney**
**What this PR does / why we need it**:

These are changes to the helm chart and documentation need to migration from the current version of the `grafana/loki` helm chart to the new one in this repo.

The current plan (as outlined in the README) is to support 3 upgrade paths:

1. From the existing `grafana/loki` chart using filesystem storage. This path will result in a single binary deployment using local filesystem storage with the new chart. This PRs includes changes to make sure this path works.
2. ~From the existing `grafana/loki` chart using object storage. This path will result in an SSD deployment with 1 read and 1 write), and has yet to be tested/documented.~ (Edit: I no longer think we need to support this use case)
3. From the existing `grafana/loki-simple-scalable` chart using object storage. This path will result in an SSD reployment with 3 read and 3 write replicas, and has yet to be tested/documented.

By clarifying the three modes, and that single binary will only support a single, filesystem backed instance, it made possible some refactors. For example, the user no longer needs to specify a `deploymentMode`, and we were able to reorganize the templates a bit into a flatter folder structure.

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: production/helm/loki/README.md [hunk 1]
```diff
@@ -25,6 +25,119 @@ Add the following repo to use the chart:
...
+
+  1. Upgrade from `grafana/loki` using local `filesystem` storage
+  1. Upgrade from `grafana/loki` using a cloud based object storage such as S3 or GCS, or an api compatible equivilent like MinIO.
+  1. Upgrade from `grafana/loki-simple-scalable` using a cloud based object storage such as S3 or GCS, or an api compatible equivilent like MinIO.
+
+### Upgrading from `grafana/loki`
+
+#### Upgrading from `filesystem` storage
+
+When Loki is backed by `filesystem` storage, we assume a single instance of Loki that is not HA. As a result, this upgrade method will involve downtime. The upgrade will involve deleting the previously deployed loki stateful set, the running the `helm upgrade` which will create the new one with the same name, which should attach to the existing PVC or ephemeral storage, thus preserving you data. Will still highly recommend backing up all data before conducting the upgrade.
```
Comment (jeschkies) [inline production/helm/loki/README.md:44]:
```suggestion
When Loki is backed by `filesystem` storage, we assume a single instance of Loki that is not HA. As a result, this upgrade method will involve downtime. The upgrade will involve deleting the previously deployed Loki stateful set, running the `helm upgrade` which will create the new one with the same name, which should attach to the existing PVC or ephemeral storage, thus preserving you data. Will still recommend backing up all data before conducting the upgrade.
```

File: production/helm/loki/README.md [hunk 2]
```diff
@@ -25,6 +25,119 @@ Add the following repo to use the chart:
...
+    type: 'filesystem'
+```
+
+Assuming you installed `grafana/loki` as `loki` in namespace `loki`, your upgrade command might look like:
+
+```console
+helm upgrade loki ~/workspace/grafana/loki/production/helm/loki --values environments/loki-migration-test/values-new.yaml --namespace loki
+```
+
+You will need to manually delete the existing stateful set for the above command to work.
```
Comment (jeschkies) [inline production/helm/loki/README.md:63]: Could you give the `kubectl` command for this?

File: production/helm/loki/README.md [hunk 3]
```diff
@@ -25,6 +25,119 @@ Add the following repo to use the chart:
...
+
+### Upgrading from `grafana/loki-simple-scalable`
+
+## Configuration
+
+By default, this chart configures Loki to run `read` and `write` targets in a scalable architecture
+designed to work with object storage. This configuration has been tested to work with [boltdb-shipper](https://grafana.com/docs/loki/latest/operations/storage/boltdb-shipper/)
+and [memberlist](https://grafana.com/docs/loki/latest/configuration/#memberlist_config) while other storage and discovery options should work as well.
```
Comment (jeschkies) [inline production/helm/loki/README.md:71]: Not sure this is worth mentioning here.

File: production/helm/loki/README.md [hunk 4]
```diff
@@ -25,6 +25,119 @@ Add the following repo to use the chart:
...
+
+If this chart is configured to run with a filesystem backend, it will assume you only want a single instance of Loki deployed
```
Comment (jeschkies) [inline production/helm/loki/README.md:73]:
```suggestion
If this chart is configured to run with a `filesystem` backend, it will assume you only want a single instance of Loki deployed
```

File: production/helm/loki/README.md [hunk 5]
```diff
@@ -25,6 +25,119 @@ Add the following repo to use the chart:
...
+in "Single Binary" mode.
```
Comment (jeschkies) [inline production/helm/loki/README.md:74]:
```suggestion
in "Single Binary" mode. See "Upgrading from `filesystem` storage" instead.
```

File: production/helm/loki/README.md [hunk 6]
```diff
@@ -25,6 +25,119 @@ Add the following repo to use the chart:
...
+
+However, the chart does not support setting up Consul or Etcd for discovery,
+and it is not intended to support these going forward.
+They would have to be set up separately.
+Instead, memberlist can be used which does not require a separate key/value store.
+The chart creates a headless service for the memberlist which read and write nodes are part of.
```
Comment (jeschkies) [inline production/helm/loki/README.md:80]: Not sure how this helps me upgrading.

File: production/helm/loki/templates/_helpers.tpl [hunk 7]
```diff
@@ -6,12 +6,63 @@ Expand the name of the chart.
 {{- coalesce .Values.nameOverride $default | trunc 63 | trimSuffix "-" }}
 {{- end }}
 
+{{/*
+singleBinary fullname
+*/}}
```
Comment (jeschkies) [inline production/helm/loki/templates/_helpers.tpl:11]: What do these helpers do?

File: production/helm/loki/templates/service-headless.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    - name: http-metrics
+      port: 3100
+      targetPort: http-metrics
+      protocol: TCP
+  selector:
+    {{- if eq (include "loki.requiresMultiStageUpgrade" . ) "true" }}
+    {{- include "loki.legacy.singleBinarySelectorLabels" . | nindent 4 }}
+    {{- else -}}
+    {{- include "loki.selectorLabels" . | nindent 4 }}
+    {{- end }}
```
Comment (jeschkies) [inline production/helm/loki/templates/service-headless.yaml:22]: Why is this required?

**Review by jeschkies — COMMENTED** [review]
I have a few questions below. Also, how did you run the upgrade test?

File: production/helm/loki/templates/_helpers.tpl (hunk 7 above)
Comment (trevorwhitney) [reply production/helm/loki/templates/_helpers.tpl:11]: the way the `loki.name` and `loki.fullname` work between the existing `grafana/loki` and `grafana/loki-simple-scalable` charts differs a bit. To enable the migration path where the new chart will replace the stateful set created by `grafana/loki` we needed to bring in this function from `grafana/loki` to make sure they had the same name.

The helpers below `isSingleBinary` and `isScalable` determine which set of templates to deploy based on whether or not object storage was requested.

File: production/helm/loki/README.md.gotmpl [hunk 9]
```diff
@@ -167,8 +101,37 @@ This chart includes dashboards for monitoring Loki. These are a work in progress
...
+        rules:
+          - record: job:loki_request_duration_seconds_bucket:sum_rate
+            expr: sum(rate(loki_request_duration_seconds_bucket[1m])) by (le, job)
+          - record: job_route:loki_request_duration_seconds_bucket:sum_rate
+            expr: sum(rate(loki_request_duration_seconds_bucket[1m])) by (le, job, route)
+          - record: node_namespace_pod_container:container_cpu_usage_seconds_total:sum_rate
+            expr: sum(rate(container_cpu_usage_seconds_total[1m])) by (node, namespace, pod, container)
+```
+
+{{ template "chart.valuesSection" . }}
```
Comment (trevorwhitney) [inline production/helm/loki/README.md.gotmpl:137]: values section moved to the bottom of README

File: production/helm/loki/README.md.gotmpl [hunk 10]
```diff
@@ -16,140 +16,74 @@ Add the following repo to use the chart:
...
-By default and inspired by Grafana's [Tanka setup](https://github.com/grafana/loki/tree/master/production/ksonnet/loki), the chart installs the gateway component which is an NGINX that exposes Loki's API
-and automatically proxies requests to the correct Loki components (read or write).
+By default and inspired by Grafana's [Tanka setup](https://github.com/grafana/loki/tree/master/production/ksonnet/loki), the chart
+installs the gateway component which is an NGINX that exposes Loki's API and automatically proxies requests to the correct
+Loki components (read or write, or single instance in the case of filesystem storage).
 The gateway must be enabled if an Ingress is required, since the Ingress exposes the gateway only.
 If the gateway is enabled, Grafana and log shipping agents, such as Promtail, should be configured to use the gateway.
 If NetworkPolicies are enabled, they are more restrictive if the gateway is enabled.
 
-## Metrics
```
Comment (trevorwhitney) [inline production/helm/loki/README.md.gotmpl:128]: Metrics section refactored into the monitoring section

File: production/helm/loki/README.md.gotmpl [hunk 11]
```diff
@@ -16,140 +16,74 @@ Add the following repo to use the chart:
...
-The chart supports the components shown in the following table.
-Ingester, distributor, querier, and query-frontend are always installed.
-The other components are optional.
-
-| Component | Optional | Enabled by default |
-| --- | --- | --- |
-| gateway |  ✅ |  ✅ |
-| write |  ❎ | n/a |
-| read |  ❎ | n/a |
+## Upgrading from v2.x
```
Comment (trevorwhitney) [inline production/helm/loki/README.md.gotmpl:19]: Added upgrade/migration documentation. Currently only the first approach is spelled out as that's the only one tested in this PR.

File: production/helm/loki/templates/gateway/configmap-gateway.yaml [hunk 12]
```diff
@@ -1,4 +1,4 @@
-{{- $isSimpleScalable := eq (include "loki.deployment.isSimpleScalable" .) "true" -}}
+{{- $isSimpleScalable := eq (include "loki.deployment.isScalable" .) "true" -}}
```
Comment (trevorwhitney) [inline production/helm/loki/templates/gateway/configmap-gateway.yaml:1]: using this opportunity to get away from the terms "Simple Scalable". This has confused community members who figured "Simple" meant not production ready.

File: production/helm/loki/templates/_helpers.tpl [hunk 13]
```diff
@@ -240,17 +270,13 @@ Create the service endpoint including port for MinIO.
...
 {{- end -}}
 
-{{/*
-Return if deployment mode is single binary
-*/}}
-{{- define "loki.deployment.isSingleBinary" -}}
-  {{- eq .Values.loki.deploymentMode "single-binary" }}
+{{/* Configure the correct name for the memberlist service */}}
+{{- define "loki.memberlist" -}}
+{{ include "loki.name" . }}-memberlist
```
Comment (trevorwhitney) [inline production/helm/loki/templates/_helpers.tpl:280]: This is a future proofing. By moving this into a function, it will make us easier to support migrations where we honor an existing memberlist DNS to enable the new deployment to join the same ring as an existing deployment. Worst case we don't need that and this remains a one line function, which I think is worth it.

File: production/helm/loki/README.md (hunk 2 above)
Comment (trevorwhitney) [reply production/helm/loki/README.md:62]: the assumption is you are using `helm`, thus there is no `kubectl` command. If a user want to run `helm template` to `kubectl` the raw yaml themselves I would consider that an advanced technique we don't need to document.

**Comment by trevorwhitney** [comment]
@jeschkies for the upgrade test, I first do `helm install loki grafana/loki --values values-old.yaml --namespace loki`, then after it's installed and stabalized, I run `helm upgrade loki ${PATH_TO_LOKI_REPO}/production/helm/loki --values values-new.yaml --namespace loki`.

`values-old.yaml`:
```yaml
---
auth_enabled: false
persistence:
  enabled: true
```

`values-new.yaml`:
```yaml
---
loki:
  commonConfig:
    replication_factor: 1
  storage:
    type: "filesystem"
```

File: production/helm/loki/templates/service-headless.yaml (hunk 8 above)
Comment (trevorwhitney) [reply production/helm/loki/templates/service-headless.yaml]: That's a good question and I've decided it's not, see my refactor.

I do think we'll need something like this when we figure out how to do a zero downtime migration, but I only think we need that for people migrating from distributed or enterprise SSD charts. Since this PR only address the migration from filesystem backed single binary, I think it's safe to assume HA is not required.

The idea here is that if we want to be able to spin up the new cluster along-side an existing cluster to do a multi-stage, zero downtime deployment, we'll need the new deployment to include the legacy label set, and we'll need to new services to use the legacy label selector so it captures both new and old pods. However, like I mentioned, not needed yet.

File: production/helm/loki/README.md (hunk 6 above)
Comment (trevorwhitney) [reply production/helm/loki/README.md:79]: hmm, yeah, let's just nuke it

File: production/helm/loki/templates/service-headless.yaml (hunk 8 above)
Comment (jeschkies) [reply production/helm/loki/templates/service-headless.yaml]: Thanks for explaining!

**Review by jeschkies — COMMENTED** [review]
I'd approve but the PR states that the migration is tested. Are you planning to add tests or will this be done in a follow up?

**Review by jeschkies — APPROVED** [review]

File: production/helm/loki/README.md (hunk 2 above)
Comment (trevorwhitney) [reply production/helm/loki/README.md:62]: ahh, yes, haha, and I suppose hit `<ctrl-d>` in `k9s` isn't appropriate, even though it's what I always use now so I've forgotten the `kubectl` command

File: production/helm/loki/README.md (hunk 5 above)
Comment (trevorwhitney) [reply production/helm/loki/README.md:73]: I added this, just in a slightly different way
